java/src/test/java/actorworkflow/CustomerLoyaltyTest.java

```diff
@@ -0,0 +1,176 @@
+                                .setWorkflowId(Shared.WORKFLOW_ID_FORMAT.formatted(guest.customerId()))
+                                .build());
+
+        testWorkflowRule
+                .getTestEnvironment()
+                .registerDelayedCallback(Duration.ofSeconds(1),
+                        () -> assertEquals(child.getStatus(), StatusTier.STATUS_TIERS.get(3)));
+
+        testWorkflowRule.getTestEnvironment().shutdown();
+    }
+
+    @Test
+    public void testAddGuestTwice() {
+        CustomerLoyaltyActivities activities = mock(CustomerLoyaltyActivities.class);
+        testWorkflowRule.getWorker().registerActivitiesImplementations(activities);
+        testWorkflowRule.getTestEnvironment().start();
+
+        // Get a workflow stub using the same task queue the worker uses.
+        WorkflowOptions workflowOptions =
+                WorkflowOptions.newBuilder()
+                        .setTaskQueue(testWorkflowRule.getTaskQueue())
+                        .setWorkflowId(Shared.WORKFLOW_ID_FORMAT.formatted("host"))
+                        .setWorkflowIdReusePolicy(
+                                WorkflowIdReusePolicy.WORKFLOW_ID_REUSE_POLICY_REJECT_DUPLICATE)
+                        .build();
+        CustomerLoyaltyWorkflow workflow =
+                testWorkflowRule
+                        .getWorkflowClient()
+                        .newWorkflowStub(CustomerLoyaltyWorkflow.class, workflowOptions);
+
+        var customer = new Customer("host", "", 0, StatusTier.STATUS_TIERS.get(4), new ArrayList<>());
+        WorkflowClient.start(workflow::customerLoyalty, customer);
+
+        int order = 0;
+        var guest = new Customer("guest");
+        testWorkflowRule.getTestEnvironment()
+                .registerDelayedCallback(Duration.ofSeconds(order++),
+                        () -> workflow.inviteGuest(guest));
+
+        testWorkflowRule.getTestEnvironment()
+                .registerDelayedCallback(Duration.ofSeconds(order++),
+                        () -> workflow.inviteGuest(guest));
+
+        testWorkflowRule.getTestEnvironment()
+                .registerDelayedCallback(Duration.ofSeconds(order++),
+                        workflow::cancelAccount);
+
+        CustomerLoyaltyWorkflow child = testWorkflowRule
+                .getWorkflowClient()
+                .newWorkflowStub(CustomerLoyaltyWorkflow.class,
+                        WorkflowOptions.newBuilder()
+                                .setTaskQueue(testWorkflowRule.getTaskQueue())
+                                .setWorkflowId(Shared.WORKFLOW_ID_FORMAT.formatted(guest.customerId()))
+                                .build());
+
+        testWorkflowRule.getTestEnvironment().registerDelayedCallback(Duration.ofSeconds(order++), () -> {
+            // "start" the workflow, to make sure we have the current execution, but expect it to throw
+            try {
+                WorkflowClient.start(child::customerLoyalty, guest);
```

Comment:
not clear on why you start child here? i think test wants to inviteGuest starting multiple children?

Reply:
Without this, I found that the `child.cancelAccount();` signal call just after this always failed with `java.lang.IllegalStateException: Null workflowId. Was workflow started?` Adding a call to start it makes it actually succeed, and I couldn't find another way to get that to work.

Reply:
ok no problem. can try to look into this bit later if thats ok